File: linux/arm/compiled/go/interpolation/interpolation_naive.go
package main

import "fmt"

// Naive Interpolation Search
func interpolationNaive(arr []int, x int) int {
	low, high := 0, len(arr)-1

	for low <= high && x >= arr[low] && x <= arr[high] {
		if arr[high] == arr[low] {
			if arr[low] == x {
				return low
			}
			break
		}

		pos := low + ((high - low) * (x - arr[low])) / (arr[high] - arr[low])

		if arr[pos] == x {
			return pos
		} else if arr[pos] < x {
			low = pos + 1
		} else {
			high = pos - 1
		}
	}

	return -1
}

func main() {
	arr := []int{2, 4, 6, 8, 10, 12, 14, 16, 18}
	target := 14

	idx := interpolationNaive(arr, target)

	if idx != -1 {
		fmt.Printf("Found %d at index %d using naive version\n", target, idx)
	} else {
		fmt.Printf("%d not found using naive version\n", target)
	}
}

File: linux/arm/interpreted/ruby/interpolation/interpolation_naive.rb
def interpolation_search_naive(arr, target)
  low = 0
  high = arr.length - 1

  while low <= high && target >= arr[low] && target <= arr[high]
    # Avoid division by zero
    return -1 if arr[high] == arr[low]

    pos = low + ((target - arr[low]) * (high - low) / (arr[high] - arr[low]))

    return pos if arr[pos] == target

    if arr[pos] < target
      low = pos + 1
    else
      high = pos - 1
    end
  end

  -1
end

# Test
arr = [2, 4, 6, 8, 10, 12, 14, 16, 18]
target = 14
index = interpolation_search_naive(arr, target)
puts "Index: #{index}"
